### ISSUE TITLE
Our changelog strategy is a pit trap

### ISSUE THREAD

**Issue by rhendric** [purescript/purescript#4130]
Our current approach to keeping a changelog is to add entries in an ‘Unreleased’ section, and then rename that section when there is a release. This means that any PR that spans a release needs to be updated every time there is a release, lest it be merged into the renamed section instead of the current ‘Unreleased’ section. There's no check to make sure that such a merge doesn't happen other than the vigilance of maintainers, and it's extra work for PR authors for the PRs that span at least one release. Personally, I think it's frustrating enough just staving off bitrot for PRs that are over a year old; moving the changelog entry around is simpler but a little more demoralizing because it's a self-created problem.

My ideal, as I've said before, would be to generate the changelog from commit messages (by strictly following some [commit message convention](https://www.conventionalcommits.org/en/v1.0.0/)), but the feedback on that, IIRC, was that we want to write our changelogs for a different audience than we want to write our commit messages for, and while I still don't agree I can respect that I'm in the minority. But that aside, there are surely other things we can do to reduce the maintenance burden on long PRs.

For example, we could keep a `CHANGELOG.md` for released changes, and a `CHANGELOG.unreleased.md` for unreleased changes, and prepend the latter into the former on release. That would keep PRs from potentially merging into the wrong part of the changelog.

Better still, we could do the above but instead of one `CHANGELOG.unreleased.md`, we could keep a folder full of snippets, one for each PR, named by category and PR number (`feature-1234.md`, `fix-1235.md`, `breaking-1236.md`). Then, not only would the above problem be solved, but we would also solve the related problem that PRs will conflict with each other if they both want to append to the same place in the changelog.

I recognize that the manual changelog was introduced in part to reduce the burden during releases; I don't think these proposals move the needle too far back in the other direction—I'll volunteer to write the script to merge changelog fragments into the released changelog, if that helps—but what do others think?

**Comment by JordanMartinez** [comment]
I think this sounds reasonable.

**Comment by hdgarrood** [comment]
Yes, this is painful. Keeping unreleased entries in either a separate file or as snippets in a directory sounds good to me!

**Comment by milesfrain** [comment]
> generate the changelog from commit messages (by strictly following some [commit message convention](https://www.conventionalcommits.org/en/v1.0.0/)), but the feedback on that, IIRC, was that we want to write our changelogs for a different audience than we want to write our commit messages for

This seems like a good idea. If the goal is to write for separate audiences, couldn't the flagged commit message notes be written for the current changelog audience? Additional commit notes (not intended for the changelog audience) would be omitted by the changelog generator tool.

**Comment by hdgarrood** [comment]
:-1: for two reasons - firstly, they are for different audiences and coming up with a scheme for working around that seems like it will be more hassle than just having changelog entries in a file like we do now, and secondly because it would be very difficult to amend changelog entries after a commit has been merged if they are part of the commit message, but that’s something which happens regularly (and should be expected to happen regularly).

**Comment by f-f** [comment]
Having a complicated system just to keep Changelogs tidy is really not worth it IMHO:
- it's more code that needs to be maintained
- it needs an explainer longer than "just add a line in a file"
- is shuffling changelog entries really an issue when rebasing a PR? In my experience the annoying part of doing that is literally everything else than the Changelog
- the person doing a release will have to open a PR anyways to bump the version number, and having a look at the Changelog to see if it's in order takes literally 30 seconds

On the other hand, this is just how I handle the Changelog in Spago. I don't really deal with this Changelog so ultimately feel free to go with whatever system is most convenient for you, but note that it might introduce friction from new contributors.

**Comment by rhendric** [comment]
@f-f

I can't contest your first two points. However...

>   * is shuffling changelog entries really an issue when rebasing a PR? In my experience the annoying part of doing that is literally everything else than the Changelog

*Remembering to do it* is the annoying part. Patrolling the open PRs, looking for which ones need to be rebased not because there's a Git-detectable merge conflict, but because they span a version and now a merge would put the changelog entry in the wrong section, is the annoying part. Keeping PRs like #4044 open even when they had enough approvals, a changelog entry written, and no merge conflicts, just because this bit of manual and purely procedural work hadn't been executed, is the annoying part. (Well, *now* #4044 has a merge conflict. But it didn't when I commented on it!) Babysitting PRs that stay open for months or years and would only need one or two rebases, but instead need one practically every time another PR is merged because of changelog conflicts, is the annoying part.

Maybe Spago just merges its PRs faster? This is all exacerbated by how long our PRs stay open and how many of them we usually have in flight at a time. But this has already bitten us once as demonstrated by #4075, and I've been playing watchdog ever since to try to keep that from happening again, and I don't think it's a good use of my time—we should either automate the watchdogging, or automate the changelog appending, and the latter saves us more work.

I know to some extent this is a me problem: nobody else asked for a volunteer to prevent changelog accidents. Maybe the occasional #4075 is an acceptable cost; maybe it's okay if it takes some extra time and attention to keep one's PRs in a constantly reviewable and mergeable state for longer than six weeks. I'm temperamentally disinclined to accept that quite so easily (not when there's *optimization* to be done!), but if I must, I must.

>   * the person doing a release will have to open a PR anyways to bump the version number, and having a look at the Changelog to see if it's in order takes literally 30 seconds

The way we currently do things, properly seeing if the changelog is in order entails scanning the previous version, or several versions, to make sure that an older PR didn't slip an entry into the wrong section because it was merged a few releases after it was written. I think having a sufficiently detailed working memory of which PRs were merged before and after which releases, or cross-referencing every PR and its merge date to check, is an unreasonable burden to place on releasers and would take considerably more than 30 seconds. But catching it at PR review time is also burdensome, because the GitHub review diff will show the entry as going into the Unreleased section, even though the merge will insert it into the section that *was* named Unreleased but *now* is named 0.14.1. And GitHub won't necessarily tell you that there's a merge conflict or anything else wrong with the PR. This is why I call it a pit trap; it'll get you without any warning.

**Comment by JordanMartinez** [comment]
@f-f

> it's more code that needs to be maintained

I think this is true, but maybe not that relevant at first thought. First, I expect that we won't need to do many changes to it once it has been written (famous last words). Second, let's say this approach is tried and we conclude that it wasn't the right way forward. We can drop the code and revert back to our current approach.

> it needs an explainer longer than "just add a line in a file"

Also true, but I don't feel like explaining how to do this is much harder. Plus, we tend not to have many new contributors who might get confused by this. So, I feel like the pros outweigh the cons. And again, if this approach was tried and found to not be the right fit, we could stop using this approach.

**Comment by f-f** [comment]
@rhendric 

> Maybe Spago just merges its PRs faster? This is all exacerbated by how long our PRs stay open and how many of them we usually have in flight at a time

YES! I am very adamant about merging things as quickly as possible into Spago's trunk, to avoid this kind of busywork. Of course it's not possible to avoid this issue in principle, and it does get naturally worse with scale (amount of code/contributors/changes).
But I firmly believe that the root of the issue is not "our changelog strategy is wrong" but really "patches should not be in flight for long". At the time of writing there are over 30 PRs open, with the oldest "ready to merge" (#3720) being almost 2 years old (!!!).

I might be the only one that feels that this is sort of unacceptable, but if I'm not then our collective time would definitely be better spent trying to figure out how to tackle the (entirely _social_) issue of getting things merged ASAP rather than coming up with a _technological_ solution for a symptom.

If it's about assurance or uncertainty that things will break then we just need more tests until se feel sure about a patch not breaking stuff - one of my refrains is "if CI passes then it's usually good by me".
If it's about having eyes/reviews on the patches then let's change the merge policy (I'm a fan of [dhall's policy](https://github.com/dhall-lang/dhall-lang/blob/master/.github/CONTRIBUTING.md#how-do-changes-get-approved), where a contributor can merge things in after a week if no one looked at them - that is, it's "default open" rather than our "default closed").
Whatever the issue is, I feel like we should be talking about it, especially [given the latest discussion about this on Discourse](https://discourse.purescript.org/t/release-prioritisation-and-cadence/2443).

**Comment by hdgarrood** [comment]
It's very easy to say that the real problem is how long compiler PRs are in flight for and that we should merge them faster. I totally agree that it is bad that compiler PRs stay open for as long as they do. At the same time, I think you are missing important context because you do not maintain the compiler. I feel that compiler PRs require the most scrutiny of any component of the entire ecosystem by a huge margin, and this is because changes to the compiler have far greater capacity to break things downstream or generate huge amounts of additional work for maintainers if they aren't done really carefully.

Saying you can just revert things is not a solution in my mind, especially for new features; as a new contributor, being told "hey, sorry, this cool feature which you worked really hard on to get merged is being reverted because we found some edge cases where it misbehaves" is probably one of the worst experiences you could have. Of course, if the feature has made it into a release before we find out about issues that it has, reverting it becomes even harder because there will now be people who rely on the feature in their code too.

The first example of this which comes to mind is the Coercible PR. Despite at least three rounds of careful reviews, and after having been open for something like a year, it was merged, and then later on we found that it was totally broken in something like 10 different ways, requiring huge amounts of follow-up work to address. At this point, reverting the change would have been really costly because of other work which went in in the meantime. Also, it was a feature that we actually wanted, and if we had reverted it, there's a good chance it never would have gotten fixed and re-merged.

I think the task of getting our test suite to a point where passing tests is a good indication that the PR is mergeable is _herculean_. And in any case, the question of what the tests are actually testing is usually more relevant than whether they pass or not, and that's just not automatable.

To be honest, this comment was a little aggravating to read. I think "avoid technological solutions to social problems" is a good idea but it also has a lot of capacity to lead people astray when applied incorrectly, and I think this probably is one of these cases. The reason I found this comment aggravating is that when you say that the current state of affairs is unacceptable, I can't really see any other way of reading this than as a demand for compiler maintainers to spend more time reviewing PRs. If there's a social problem here, it's that maintaining an open source compiler in which PRs get fast feedback requires an order of magnitude more maintainer hours than we can currently offer.

It might be instructive to look at stats like numbers of open PRs and average time to merge for other open source compilers which do have proper funding structures, but I really think that 90% of this is that you just can't apply expectations from an average open source project to compilers, because compilers are just in a different class in terms of maintenance difficulty.

**Comment by JordanMartinez** [comment]
> I might be the only one that feels that this is sort of unacceptable, but if I'm not then our collective time would definitely be better spent trying to figure out how to tackle the (entirely social) issue of getting things merged ASAP rather than coming up with a technological solution for a symptom.

Just a few thoughts.

To get a PR merged requires two things: CI to build either two approvals from a core team member if the contributor is not one or an approval from one other core team member if the contributor is a member. CI hasn't been a problem. 

Let's look at each of the 30 PRs to imagine why they haven't been merged yet:
-

**Comment by JordanMartinez** [comment]
(Sorry about that. Hit the wrong button)

**Comment by f-f** [comment]
@hdgarrood 

It sounds like you might have taken this personally, and if that was the case I'm sorry that happened.
Demanding open source work from people is possibly one of the last things I'd do because (1) we're literally on the same boat and (2) if I need/want something done I just, uh, get up and do it.
It's really unfortunate you just picked this interpretation instead of asking a clarifying question about the intention behind my comment, so in the interest of future constructive conversations I'm going to ask you to, like, assume people are trying to help instead of assuming bad faith? (yes, "demanding open source work" is "bad" in my book)

What I meant to do above was merely to offer a perspective and a point of convergence to focus collective effort - in which I consider myself included of course, otherwise what else I'm doing over here? - that might be instead misdirected (i.e. far from optimal on several fronts).
I think this is a perfectly reasonable perspective to have, but as I stated above I might actually just be the only one feeling this way, in which case you can choose to just ignore me and move on with your day, instead of getting aggravated about my perspective (I encourage this!)

Re "I don't maintain the compiler": I ship software for money to people since 10 years, I know what it means to ship broken stuff and so on, I well understand that the compiler is a sensitive piece of software. However, part of my day job includes working exactly on removing the collective (and irrational IMO) fear attached to "what if we ship broken stuff?", by just building more assurance until we're confident that CI will catch most of our mistakes.
There's obviously a point of diminishing returns, but we're faaar from there in this case.

Re improving tests: I strongly disagree with the herculeannes of the effort. I have a couple of cheap examples that are easy to implement and would easily catch regressions:
- having code coverage and demanding every patch to improve coverage - this is really easy to do with `stack` and would uncover immediately undertested new code
- compile the whole package set in CI - it only takes 5 mins, and would have for example caught the regression that delayed 0.14.3

You bring up the Coercible PR as a negative example, but I think that was very well managed! Thing got merged in, it was found broken in some spots, things got fixed up, that sounds..fine to me? Not everything can be reverted, and that's ok; what matters to me in this discussion is to not let good patches go stale.

Re financing development: I'm of couse on board with helping with setting that up if we think it would be a worth effort. I think it would be very worth but it requires coordination and I cannot do it alone.

**Comment by JordanMartinez** [comment]
> It's really unfortunate you just picked this interpretation instead of asking a clarifying question about the intention behind my comment, so in the interest of future constructive conversations I'm going to ask you to, like, assume people are trying to help instead of assuming bad faith?

For what it's worth, I also read your response [above](https://github.com/purescript/purescript/issues/4130#issuecomment-876388581) as a bit aggressive, which initially surprised me. I think it was because you said "unacceptable." Regardless, it's clear that you didn't say that in bad faith and that my interpretation was just wrong.

Getting back to the point, I am currently looking at each of the PRs that are open and analyzing why they are in the state that they are in. It might give a better summary than just "30 or so PRs are open". For example, I believe 4 or so are currently blocked until v0.14.3 is officially announced. Another two or so likely should have been closed but haven't been yet. Another two should probably be merged, but just haven't been merged yet. I haven't looked at the others yet or categorized them.

Still, I think a large part why we can get into this situation of having many opens PRs is because of posts like the Discourse one you linked to above. I feel like we are often reactionary, responding to the next biggest issue (and sometimes, complaint) and then sometimes forgetting what issues we were tackling before such a post/issue/PR was raised/made. In other cases, there's just so many updates in a short period that it's hard to remember what we were going to do two weeks ago.

**Comment by rhendric** [comment]
This larger conversation is great, I hope we keep having it, and I have things to say about all the responses here... but I think it speaks to one of PureScript's explicit project values, ‘prefer to move slowly and consider decisions carefully,’ and as such I don't expect it to resolve very soon.

Getting back to the *point*, I don't think the changelog problems I'm pointing out are a symptom of a social issue. I think they're caused by a technological issue that can be exacerbated or alleviated by less or more bias to action, but can be defined and solved independently. The technological issue is a mismatch between what Git and GitHub thinks a change to the changelog is, and how we actually use our changelog. Renaming the Unreleased section to a version number should, conceptually, ‘take out a lock’ on the entire section, so that ‘transactions’ that add entries to that section can fail. AFAIK there's no way to do that in Git (short of editing every line somehow, or moving the content between files), and so I think we should adapt our usage pattern to our tool. The only purpose of the technology in #4132 is to make adapting to the new usage pattern as easy as possible for contributors and releasers, while implementing an appropriately granular locking strategy—the fundamental point in this issue and that PR is to address the mismatch, the nontransactional usage of a conceptually transactional resource which predictably results in inconsistent state.

Addressing this technological issue with a social solution (reconsidering our project values) is IMO the approach that involves treating the symptom instead of the cause, even if those project values should be reconsidered.

And to your points about bias to action, @f-f, there are several open PRs right now that I would love to shepherd closer to merging by advising that the changelog entry be updated or added, but it'd be unreasonable to do that before we have consensus on what the changelog format is going to be going forward. Can we please focus first on getting that consensus?

**Comment by f-f** [comment]
> I think it was because you said "unacceptable."

@JordanMartinez I see, that makes sense. I think part of this is the slightly different meaning this might have in my native language, apologies if that sounded weird. The overall concept that I wanted to convey with that is something like "we should not settle for this and we deserve better"

> Can we please focus first on getting that consensus?

@rhendric sure, as I mentioned above you can feel free to ignore my tangent and move forward with whatever is best for y'all

**Comment by milesfrain** [comment]
> AFAIK there's no way to do that in Git (short of editing every line somehow)

We could add an "unreleased" prefix to each bullet in the unreleased section, which we then strip out once moved to an actual release. That would at least prevent bad merges, although I like how #4132 adds other conveniences to avoid merge conflicts. I'm surprised these issues (and workarounds) haven't been discussed already in the [keep-a-changelog repo](https://github.com/olivierlacan/keep-a-changelog/issues). Seems like any large project would eventually encounter these changelog annoyances.

**Comment by rhendric** [comment]
First relevant thing I found in that repo is olivierlacan/keep-a-changelog#345, which looks like a tool that operates in the same space as the script I wrote for #4132.

**Comment by JordanMartinez** [comment]
> I see, that makes sense. I think part of this is the slightly different meaning this might have in my native language, apologies if that sounded weird. The overall concept that I wanted to convey with that is something like "we should not settle for this and we deserve better" (note that this is not judgemental and no one is at fault here, organic evolution is a normal state of software development)

That (that you likely think in one language that you then translate to English when conversing with us and by default such translations are not always a 1-to-1 translation) makes sense! And yes, I think "not settling" and finding a better path forward for anything is always desirable.

**Comment by hdgarrood** [comment]
> It's really unfortunate you just picked this interpretation instead of asking a clarifying question about the intention behind my comment, so in the interest of future constructive conversations I'm going to ask you to, like, assume people are trying to help instead of assuming bad faith? (yes, "demanding open source work" is "bad" in my book)

I did not assume bad faith; at no point did I assume that you were saying anything which was not a sincerely held opinion. If what you were intending to say was "we should arrange things so that it is easier to get PRs merged faster," then I have no trouble with that point of view. However, I'm not going to be made to feel bad for getting aggravated by what you actually wrote:

> But I firmly believe that the root of the issue is not "our changelog strategy is wrong" but really "patches should not be in flight for long". At the time of writing there are over 30 PRs open, with the oldest "ready to merge" (#3720) being almost 2 years old (!!!). I might be the only one that feels that this is sort of unacceptable

When you say it is unacceptable that patches are in flight for this long, what you're saying is that this is a state of affairs that _needs_ to be addressed, even if addressing that comes with a lot of cost. Even if you might not have realised this, an unavoidable component of that cost (in my view) is "maintainers need to put more hours in", which I think is reasonable for me to find aggravating, as is your saying we "just" need more tests.

> Re improving tests: I strongly disagree with the herculeannes of the effort. [...]

Those two things would probably help, although I think it's worth noting that we previously had both of them in place, and ended up discarding them because we decided (perhaps incorrectly) that they were too much effort to keep running properly. In any case: would they be sufficient for me to start saying "tests are green so we can merge?" Absolutely not. There is no substitute for having a maintainer carefully review, and think through the implications of the change and how it has been implemented.
